### PR TITLE
Controls-Tech-DBW Integration - `2-15-test-day`

### PR DESCRIPTION
**Pull request by giuseppequaratino**
Merge in encoder_odom and techbus ros nodes into dev. Also remove some unused ros packages in an effort to start cleaning the tech stack

### CODE REVIEW

File: can/can.yml [hunk 1]
```diff
@@ -289,6 +289,7 @@ nodes:
         id: 0x21
         cycletime: 100
 
+        # GO INTO DBC search for angleCmd and change @1+ to @1- for signed values
```
Comment (giuseppequaratino) [inline can/can.yml]: outdated comment

File: can/can.yml [hunk 2]
```diff
@@ -298,6 +299,20 @@ nodes:
...
+        cycletime: 100
+
+        signals:
+          - DesiredVel:
+              description: supposed linear.x output.
+              width: 32 #???? girl tf is a width
+          - Angle:
+              description: supposed angle *giggle*
+              width: 32 #probably idfk
+
```
Comment (giuseppequaratino) [inline can/can.yml]: unused message meant for testing

File: ros/docker/encoder_odom/src/encoder_arduino.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env python3
```
Comment (giuseppequaratino) [inline ros/docker/encoder_odom/src/encoder_arduino.py:1]: This file hasn't been used for testing yet. Needs to incorporate the changes from encoder_odom.py as well so probably shouldn't merge it.

File: .editorconfig [hunk 4]
```diff
@@ -1,15 +0,0 @@
-# Cooper IGVC EditorConfig
```
Comment (dmezh) [inline .editorconfig:1]: Why deleted?

File: Makefile [hunk 5]
```diff
@@ -64,5 +63,5 @@ $(INSTALL_DEPENDENCIES): $(REQUIREMENTS_TXT)
 	$(PYTHON) -m $(PIP) install --upgrade pip wheel
 	$(PYTHON) -m $(PIP) install --upgrade $(LOCAL_PYTHON_LIBS)
 	$(PYTHON) -m $(PIP) install --requirement $(REQUIREMENTS_TXT)
-	cargo install --root build/cargo --locked --git https://github.com/opencan/opencan --rev 3206e406a
+	cargo install --root build/cargo --locked --git https://github.com/opencan/opencan --rev f90d5c42
```
Comment (dmezh) [inline Makefile]: - [ ] Change version to non-hack before merge

File: ctrl/velocity-controller/ctrl.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+import vel
+
+
+class VelocityControl(threading.Thread):
+    SEND_RATE_S = 0.01
+
+    def __init__(self, *, bus: cand.client.Bus = None):
+        self._bus      = bus
+
+        self.throttle_percent = 0
```
Comment (dmezh) [inline ctrl/velocity-controller/ctrl.py]: Should these also start off as floats?

File: ctrl/velocity-controller/vel.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+import igvcutils
+import numpy as np
+
+
+class Ctrl:
+    ENCODER_TICKS_PER_ROTATION = 4000
+    WHEEL_CIRCUMFERENCE_M      = 1.899156
+    TICKS_PER_M                = WHEEL_CIRCUMFERENCE_M / ENCODER_TICKS_PER_ROTATION
+
+    ACCEL_TO_PEDAL_SLOPE_MAPPING        = 15.4
```
Comment (dmezh) [inline ctrl/velocity-controller/vel.py:13]: imo these names could be more clear; it's not apparent what `X_TO_SLOPE_MAPPING` means

File: dbw/py-steering/base.py [hunk 8]
```diff
@@ -81,14 +81,15 @@ def run(self):
...
-        self._sys_state = 'ESTOP'
-        self._bus.send(
-            'DBW_ESTOP',
-            {'DBW_src': 'NODE', 'DBW_reason': reason},
-        )
-        self._logger.warn(f'ESTOP reason: {reason}')
-        if err_msg: self._logger.error(err_msg)
-        self._logger.critical('state: ESTOP')
+        pass
+#        self._sys_state = 'ESTOP'
```
Comment (dmezh) [inline dbw/py-steering/base.py:85]: Is this staying or going?

File: dbw/py-steering/steer.py [hunk 9]
```diff
@@ -131,7 +131,7 @@ def run(self):
             self._bus.send(
                 'STEER_SteeringData',
                 {
-                    'STEER_angle':             math.radians(self._cur_angle),
```
Comment (dmezh) [inline dbw/py-steering/steer.py:134]: Change this back once we get signed support merged?

File: ros/docker-compose.yml [hunk 10]
```diff
@@ -25,16 +25,15 @@ services:
...
-    twist:
-        build: docker/cmd_to_ctrl/.
-        container_name: twist
+    pure_pursuit:
+        build: docker/pure_pursuit/.
+        container_name: pure_pursuit
         tty: true
         environment:
-            - 'ROS_HOSTNAME=navigation'
+            - 'ROS_HOSTNAME=pure_pursuit'
```
Comment (dmezh) [inline ros/docker-compose.yml:33]: :)

File: ros/docker-compose.yml [hunk 11]
```diff
@@ -168,3 +150,27 @@ services:
...
+        environment:
+            - 'ROS_HOSTNAME=encoder_odom'
+            - 'ROS_MASTER_URI=http://master:11311'
+        volumes:
+            - './docker/encoder_odom/src:/app/ros_ws/src/encoder_odom'
+        depends_on:
+            - master
+
+    techbus:
+        build: docker/techbus/.
```
Comment (dmezh) [inline ros/docker-compose.yml]:
```suggestion
        build: docker/techbus/
```

File: ros/docker/encoder_odom/Dockerfile [hunk 12]
```diff
@@ -2,17 +2,28 @@ FROM ros:noetic-ros-base
...
 
+WORKDIR /app
+
 RUN apt update && apt install -y curl
 
+RUN apt-get -y update
+
+RUN apt-get install -y python3-tf2-ros
+
+RUN apt-get install -y python3-tf-conversions
```
Comment (dmezh) [inline ros/docker/encoder_odom/Dockerfile:13]: Docker caches in between these RUN directives, meaning the `apt-get update` line could be cached and then the `apt-get install` line runs on a cached version of the apt lists. You need to run update in the same RUN directive:

```suggestion
RUN apt-get update && apt-get install -yq python3-tf2-ros python3-tf-conversions
```

File: ros/docker/encoder_odom/src/encoder_odom.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+import tf2_ros
+import tf_conversions
+import ctypes
+import math
+from math import sin, cos, pi
+
+from geometry_msgs.msg import Point, Pose, Quaternion, Vector3, Twist, TransformStamped 
+from std_msgs.msg import UInt16MultiArray, String
+from nav_msgs.msg import Odometry
+
```
Comment (dmezh) [inline ros/docker/encoder_odom/src/encoder_odom.py:13]: order imports

File: ros/docker/encoder_odom/src/encoder_odom.py [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+        self.delta_x = 0.0
+        self.delta_y = 0.0
+        self.delta_th = 0.0
+
+        self.delta_left = 0
+        self.delta_right = 0
+        self.v_left = 0.0
+        self.v_right = 0.0
+
+        #starting the node at 0 ticks might cause a problem with absolute encoders???
```
Comment (dmezh) [inline ros/docker/encoder_odom/src/encoder_odom.py:52]: yes, you should wait to publish encoder data until you see the first tick and set the initial values to it.

File: ros/docker/encoder_odom/src/encoder_odom.py [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+        t.transform.rotation.w = self.q[3]
+        
+        self.broadcaster.sendTransform(t)
+    def calc_odom(self,current_time,last_time):
+        
+        # Car Variables
+        wheel_radius = 0.3302   #13 inch ~= 0.3302 meter
+        circumfrence = 1.899156
+        enc_res = 65536
+        tick_distance = circumfrence/enc_res
```
Comment (dmezh) [inline ros/docker/encoder_odom/src/encoder_odom.py]:
```suggestion
        tick_distance = circumference/enc_res
```

File: ros/docker/encoder_odom/src/encoder_odom.py [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+    encoder_odom = Encoder_Odom() 
+
+    current_time = rospy.Time.now()
+    last_time = rospy.Time.now()
+       
+    while not rospy.is_shutdown():
+
+        current_time = rospy.Time.now()
+        
+        pub.publish("hello")
```
Comment (dmezh) [inline ros/docker/encoder_odom/src/encoder_odom.py]: remove?

File: ros/docker/pure_pursuit/src/global_ros_pp.py [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+                 #x and y value of ld
+            elif (t2 >= 0 and t2 <=1.0):
+                point = waypoints[wp] + t2*d
+                pointx = point[0]
+                pointy = point[1]
+                return pointx, pointy
+            else:
+                return 0,0
+
+    except NoIntersectionException as e:
```
Comment (dmezh) [inline ros/docker/pure_pursuit/src/global_ros_pp.py:106]: An exception is thrown and then immediately caught, so is there a point? In other words, what do you actually want to do when we have no intersection?

File: ros/docker/pure_pursuit/src/global_ros_pp.py [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+        print("Path lookahead no intersection")
+        return -999, -999
+
+# Calculates the steering angle needed to follow the arc connecting the current position and the lookahead point
+def steering_angle(lookahead_x, lookahead_y, xpos, ypos, yaw):
+    
+    angle = np.tan(yaw)
+    a = -1 * angle
+    c = angle * xpos - ypos
+    dist_x = abs(a*lookahead_x + lookahead_y + c)/math.sqrt(((a)**2)+1)
```
Comment (dmezh) [inline ros/docker/pure_pursuit/src/global_ros_pp.py:116]: Please make spacing more consistent

File: ros/docker/pure_pursuit/src/global_ros_pp.py [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+
+            ppc_cmd.data = [v_next, steer_next]
+            publisher.publish(ppc_cmd)
+            
+        # Once we get within a predefined distance of the endpoint, we want to stop accelerating
+        a_max = 0
+        v_next = decel(decel_dist, dist, v_m, a_max, subscriber.car.xpos)
+
+        ppc_cmd.data = [v_next, steer_next]
+        publisher.publish(ppc_cmd)
```
Comment (dmezh) [inline ros/docker/pure_pursuit/src/global_ros_pp.py:192]: Missing newline at end of file

File: ros/docker/pure_pursuit/src/path_gen.py [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+    for i in range (0,n_points-1): 
+        path.PoseStamped[i].pose.position.x = i/100
+        path.PoseStamped[i].pose.position.y = 10*np.sin(0.1*i/100) # Replace with whatever function you want
+        # path.PoseStamped[i].pose.position.y = np.sqrt(1+10*i)
+        # path.PoseStamped[i].pose.position.y = 0
+
+    while not rospy.is_shutdown():
+
+        pub_path.publish(path)
+        rate.sleep()
```
Comment (dmezh) [inline ros/docker/pure_pursuit/src/path_gen.py]: Missing newline at end of file

File: ros/docker/pure_pursuit/src/ppwork.py [hunk 21]
```diff
@@ -0,0 +1,233 @@
+import math
```
Comment (dmezh) [inline ros/docker/pure_pursuit/src/ppwork.py]: Which one of the pure pursuit controller files is correct?

File: ros/docker/pure_pursuit/src/path_gen.py [hunk 22]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env python3
```
Comment (dmezh) [inline ros/docker/pure_pursuit/src/path_gen.py]: Maybe a different name for this file would be good - `path_test_gen.py`?

File: ros/docker/pure_pursuit/src/ros_pp.py [hunk 23]
```diff
@@ -0,0 +1,195 @@
+#!/usr/bin/env python3
```
Comment (dmezh) [inline ros/docker/pure_pursuit/src/ros_pp.py:1]: A third pure pursuit controller file - which one is right?

File: ros/docker/techbus/Dockerfile [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+    source .venv/bin/activate && \
+    python3.9 -m pip install opencan-cand numpy msgpack cantools
+
+RUN curl -fsSL https://packages.redis.io/gpg | gpg --dearmor -o /usr/share/keyrings/redis-archive-keyring.gpg
+
+RUN echo "deb [signed-by=/usr/share/keyrings/redis-archive-keyring.gpg] https://packages.redis.io/deb $(lsb_release -cs) main" | sudo tee /etc/apt/sources.list.d/redis.list
+
+RUN apt-get update && apt-get install -qy redis
+
+
```
Comment (dmezh) [inline ros/docker/techbus/Dockerfile:24]:
```suggestion
```

File: ros/docker/techbus/Dockerfile [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+RUN apt update && apt install -y git
+
+RUN apt-get update && apt install -y python3.9 python3.9-venv python3-pip --fix-missing
+
+RUN apt update && apt install -y curl
+
+RUN python3.9 -m venv .venv && \
+    source .venv/bin/activate && \
+    python3.9 -m pip install opencan-cand numpy msgpack cantools
+
```
Comment (dmezh) [inline ros/docker/techbus/Dockerfile:17]:
```suggestion

# install redis
```

File: ros/docker/techbus/src/techbus.py [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+class ROStouCAN:
+    def __init__(self):
+        self.bus = cand.client.Bus(redis_host='redis')
+        rospy.Subscriber(PPTwist, Twist, self.callback)
+    def callback(self, msg):
+        angle = 0
+        if msg.linear.x != 0 and msg.angular.z != 0:
+            angle = -atan2(msg.angular.z * 1.8, msg.linear.x)
+        else:
+            rospy.loginfo(f"Warning: invalid steering angle combo:lin.x:{msg.linear.x}, ang.z:{msg.angular.z}")
```
Comment (dmezh) [inline ros/docker/techbus/src/techbus.py:33]:
```suggestion
            rospy.loginfo(f"Warning: invalid steering angle/linear velocity combo: lin.x:{msg.linear.x}, ang.z:{msg.angular.z}")
```

File: ros/docker/techbus/src/techbus.py [hunk 27]
```diff
@@ -0,0 +1,80 @@
...
+            angle = 0
+        self.bus.send(VelocityCAN, {
+            'DBW_linearVelocity': msg.linear.x
+        })
+        self.bus.send(SteerCAN, {
+            'STEER_angleCmd': angle
```
Comment (dmezh) [inline ros/docker/techbus/src/techbus.py]: We should make the message `DBW_steeringAngleCmd`

File: ros/docker/techbus/src/techbus.py [hunk 28]
```diff
@@ -0,0 +1,80 @@
...
+# sudo apt install python3.9
+# python3.9 -m pip install opencan-cand
+
+# start redis container
+# set up vcan on local machine 
+# sudo modprobe vcan
+# sudo ip link add dev vcan0 type vcan
+# sudo ip link set up vcan0
+# start cand on local machine candump vcan0
+# redis local host flag on cand --redis_host localhost
```
Comment (dmezh) [inline ros/docker/techbus/src/techbus.py]: Missing newline at end of file

File: ros/test-scripts/calibrate-start.sh [hunk 29]
```diff
@@ -0,0 +1,83 @@
...
+tmux rename-window 'cand'
+tmux send-keys 'cand --dev can0 --dbc ../../build/can/igvc_can.dbc'
+tmux send-keys Enter
+
+tmux split-window -v 'candump'
+
+tmux send-keys 'sudo ip link set can0 up type can bitrate 500000'
+tmux send-keys Enter
+tmux send-keys 'candump can0 | cantools decode build/can/igvc_can.dbc -s | grep Cmd'
+tmux send-keys Enter
```
Comment (dmezh) [inline ros/test-scripts/calibrate-start.sh:83]: Missing newline at end of file

File: ros/test-scripts/cand-start.sh [hunk 30]
```diff
@@ -0,0 +1,24 @@
...
+tmux rename-window 'cand'
+tmux send-keys 'cand --dev can0 --dbc ../../build/can/igvc_can.dbc'
+tmux send-keys Enter
+
+tmux split-window -v 'candump'
+
+tmux send-keys 'sudo ip link set can0 up type can bitrate 500000'
+tmux send-keys Enter
+tmux send-keys 'candump can0 | cantools decode build/can/igvc_can.dbc -s | grep Cmd'
+tmux send-keys Enter
```
Comment (dmezh) [inline ros/test-scripts/cand-start.sh:24]: Missing newline at end of file

File: startup.sh [hunk 31]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
```
Comment (dmezh) [inline startup.sh:2]: Maybe leave a comment explaining the script

**Review by dmezh — CHANGES_REQUESTED** [review]

**Comment by dmezh** [comment]
Merging this will be huge! Exciting work 😄

**Comment by dmezh** [comment]
What's the status on this?

**Review by dmezh — CHANGES_REQUESTED** [review]
Please address the outstanding comments.

**Review by dmezh — APPROVED** [review]
Approving on condition that outstanding comments get resolved later over time.

**Pull request merged**